api/src/exam-environment/routes/exam-environment.ts

```diff
@@ -0,0 +1,551 @@
+/* eslint-disable jsdoc/require-returns, jsdoc/require-param */
+import { type FastifyPluginCallbackTypebox } from '@fastify/type-provider-typebox';
+import { type FastifyInstance, type FastifyReply } from 'fastify';
+import jwt from 'jsonwebtoken';
+
+import * as schemas from '../schemas';
+import { mapErr, syncMapErr, UpdateReqType } from '../../utils';
+import { JWT_SECRET } from '../../utils/env';
+import {
+  checkAttemptAgainstGeneratedExam,
+  checkPrerequisites,
+  constructUserExam,
+  generateExam,
+  userAttemptToDatabaseAttemptQuestionSets,
+  validateAttempt
+} from '../utils/exam';
+import { ERRORS } from '../utils/errors';
+
+/**
+ * Wrapper for endpoints related to the exam environment desktop app.
+ *
+ * Requires exam environment authorization token to be validated.
+ */
+export const examEnvironmentValidatedTokenRoutes: FastifyPluginCallbackTypebox =
+  (fastify, _options, done) => {
+    // TODO: Is there any reason to delete the token without generating a new one?
+    fastify.post(
+      '/exam-environment/exam/generate',
+      {
+        schema: schemas.examEnvironmentPostExamGenerate
+      },
+      postExamGenerateHandler
+    );
+    fastify.post(
+      '/exam-environment/exam/attempt',
+      {
+        schema: schemas.examEnvironmentPostExamAttempt
+      },
+      postExamAttemptHandler
+    );
+    fastify.post(
+      '/exam-environment/screenshot',
+      {
+        schema: schemas.examEnvironmentPostScreenshot
+      },
+      postScreenshotHandler
+    );
+    done();
+  };
+
+/**
+ * Wrapper for endpoints related to the exam environment desktop app.
+ *
+ * Does not require exam environment authorization token to be validated.
+ */
+export const examEnvironmentOpenRoutes: FastifyPluginCallbackTypebox = (
+  fastify,
+  _options,
+  done
+) => {
+  fastify.post(
+    '/exam-environment/token/verify',
+    {
+      schema: schemas.examEnvironmentTokenVerify
+    },
+    tokenVerifyHandler
+  );
+  done();
+};
+
+interface JwtPayload {
+  examEnvironmentAuthorizationToken: string;
+}
+
+/**
+ * Verify an authorization token has been generated for a user.
+ *
+ * Does not require any authentication.
+ *
+ * **Note**: This has no guarantees of which user the token is for. Just that one exists in the database.
```

Comment:
Good point. I cannot think of a reason not to.

Reply:
Good point about the status codes. I would reuse the logic, though. Otherwise the verification logic could incorrectly report that it's allowed when not and vice versa.